src/hal/src/external_memory/mod.rs

```diff
@@ -0,0 +1,498 @@
+            // Without this on non unix or windows platform, this will trigger error for unreachable pattern
+            #[allow(unreachable_patterns)]
+            _ => Err("Wrong handle type and platform memory combination"),
+        }
+    }
+}
+
+#[cfg(unix)]
+impl std::convert::TryFrom<(ExternalMemoryType, Fd)> for ExternalMemory {
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, Fd)) -> Result<Self, Self::Error> {
+        ExternalMemory::try_from((tuple.0, PlatformMemory::from(tuple.1)))
+    }
+}
+
+#[cfg(windows)]
+impl std::convert::TryFrom<(ExternalMemoryType, Handle)> for ExternalMemory {
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, Handle)) -> Result<Self, Self::Error> {
+        ExternalMemory::try_from((tuple.0, PlatformMemory::from(tuple.1)))
+    }
+}
+
+impl std::convert::TryFrom<(ExternalMemoryType, Ptr)> for ExternalMemory {
+    type Error = &'static str;
+    fn try_from(tuple: (ExternalMemoryType, Ptr)) -> Result<Self, Self::Error> {
+        ExternalMemory::try_from((tuple.0, PlatformMemory::from(tuple.1)))
+    }
+}
+
+bitflags::bitflags! {
+    /// Possible usages for an image.
+    pub struct ImageUsage: u32 {
+        /// Format can be used for sampled images (SAMPLED_IMAGE and COMBINED_IMAGE_SAMPLER descriptor types).
+        const SAMPLED_IMAGE = (1 << 0);
+        /// Format can be used for storage images (STORAGE_IMAGE descriptor type).
+        const STORAGE_IMAGE = (1 << 1);
+        /// Format supports atomic operations in case it is used for storage images.
+        const STORAGE_IMAGE_ATOMIC = (1 << 2);
+        /// Format can be used for uniform texel buffers (TBOs).
+        const UNIFORM_TEXEL_BUFFER = (1 << 3);
+        /// Format can be used for storage texel buffers (IBOs).
+        const STORAGE_TEXEL_BUFFER = (1 << 4);
+        /// Format supports atomic operations in case it is used for storage texel buffers.
+        const STORAGE_TEXEL_BUFFER_ATOMIC = (1 << 5);
+        /// Format can be used for vertex buffers (VBOs).
+        const VERTEX_BUFFER = (1 << 6);
+        /// Format can be used for color attachment images.
+        const COLOR_ATTACHMENT = (1 << 7);
+        /// Format supports blending in case it is used for color attachment images.
+        const COLOR_ATTACHMENT_BLEND = (1 << 8);
+        /// Format can be used for depth/stencil attachment images.
+        const DEPTH_STENCIL_ATTACHMENT = (1 << 9);
+        /// Format can be used as the source image of blits with vkCmdBlitImage.
+        const BLIT_SRC = (1 << 10);
+        /// Format can be used as the destination image of blits with vkCmdBlitImage.
+        const BLIT_DST = (1 << 11);
+        /// Format can be filtered with VK_FILTER_LINEAR when being sampled
+        const SAMPLED_IMAGE_FILTER_LINEAR = (1 << 12);
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+#[derive(Debug)]
+/// Drm format properties
+pub struct DrmFormatProperties {
```

Comment:
Do we need to have it here? If it's a Vulkan/Linux-only structure and an entry point, can we just make them public in `gfx-backend-vulkan` instead of having them in gfx-hal?

Reply:
I could move it to the Vulkan backend if you desire, i have wrote it in hal because OpenGL also got the possibility to use this feature, even if i'm not enough familiar to implement on it's backend. I know that this feature is used by current Wayland compositors to match the OpenGL image format with the drm/kms surfaces one. The Vulkan extension also states that it is equivalent to using 2 EGL extensions:
https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VK_EXT_image_drm_format_modifier.html#_description

Reply:
Basically what we want is gfx-hal having no cfg() switches. If this is supported on Vk and Gl, then let's make it an optional feature that just happens to be linux-only. Would it work?

Reply:
`DrmFormatProperties` is now returned as part of `format::Properties` as `Vec<DrmFormatProperties>`. Cfg flags as been removed and if the platform it is not supported, the field will simply be an empty vec. This change has allowed to merge the required function into the already existing `PhysicalDevice::format_properties` so no new function on hal is required.